MNNVL support design doc (#297)
ai-dynamo/grove: docs/designs/mnnvl-design.md

# MNNVL Support Design Doc \- Phase 1

# Overview

This design document details the plan to enable the Grove operator to automatically leverage MNNVL for appropriate workloads.

## Motivation

The MNNVL support feature in Grove is guided by three core design principles:

> **Simplicity:** Deploying inference workloads with MNNVL in Kubernetes currently demands familiarity with complex concepts like ComputeDomain CRDs, ResourceClaimTemplates, and Dynamic Resource Allocation (DRA) APIs. Grove, however, must completely abstract this complexity, minimizing required configuration and hiding all underlying implementation details from the user.

Reviewer: Maybe rephrase this as:
Currently, inference workloads in Kubernetes can take advantage of NVIDIA MNNVL accelerated hardware by explicitly adding NVIDIA ComputeDomain resource claims to their workloads. With this feature, when applicable, Grove will abstract the MNNVL hardware from users by automatically creating ComputeDomain custom resources, enabling ComputeDomains support to workloads, and managing the lifecycle of the CRs.

**Portability:** The Grove Custom Resource Definitions (CRDs)—specifically PodCliqueSet, PodCliqueScalingGroup, and PodClique—are designed to be generic, excluding any direct references to MNNVL, ComputeDomain, or other NVIDIA-specific components.

> To maintain portability, all MNNVL-related configuration is kept external, residing in the operator configuration and runtime annotations. This approach allows the same PodCliqueSet manifest to be used without alteration, regardless of whether the target cluster supports MNNVL.

Author: Added a comment about it in the "Enabling the feature" section

> **Failure Resilience:** MNNVL is treated as an optimization, not a hard requirement. If encountering a failure creating the MNNVL resources, Grove does not block the workload. Instead, pods are created without the RCT reference and are scheduled using standard GPU allocation mechanisms.

Author: You are right, I renamed it "Graceful degradation."

> **Failure Resilience:** MNNVL is treated as an optimization, not a hard requirement. If encountering a failure creating the MNNVL resources, Grove does not block the workload. Instead, pods are created without the RCT reference and are scheduled using standard GPU allocation mechanisms.

Reviewer 2: Lack of next steps is a great callout, want to make sure we don't scope creep during implementation.

Based on #296, some suggestions to add in:
  - Phase 1 only supports homogeneous clusters
  - DRA-only GPU requests not supported (Phase 2, potentially needs more color around how we support "power users"
  - No per-PCLQ granularity for MNNVL toggle

### Using MNNVL in a K8S cluster

To use **MNNVL (Multi-Node NVLink)** in Kubernetes with NVIDIA’s DRA driver, you start by creating a **ComputeDomain (CD)**. The ComputeDomain represents a logical GPU fabric spanning multiple nodes and defines how inter-node NVLink/IMEX channels are allocated. In its `spec.channel`, the ComputeDomain references a **ResourceClaimTemplate**; this template describes the DRA resource class (MNNVL) that will be used to provision the interconnect. When the ComputeDomain is created, the NVIDIA DRA controller prepares the underlying GPU fabric and ensures that nodes participating in the domain can securely communicate over MNNVL.

Next, pods that want to use MNNVL simply **reference the ResourceClaimTemplate** in their pod spec. Each pod declares a `resourceClaims` entry pointing to the template name, and the container lists that claim under `resources.claims`. Kubernetes then automatically creates a **ResourceClaim per pod** from the template. These per-pod claims are handed to the NVIDIA DRA driver, which allocates the necessary MNNVL/IMEX channels for each pod within the ComputeDomain.

> As a result, multiple pods—possibly scheduled on different nodes—are joined into the same ComputeDomain and can communicate using **multi-node NVLink semantics** without manually wiring GPUs or fabric resources. The ComputeDomain handles reachability and isolation, the ResourceClaimTemplate enables automatic per-pod allocation, and the pod spec remains simple: create a CD once, then reference its template in every pod that needs MNNVL.

Author: fix

> # Goals

Reviewer 2: Agreed

The following key goals, derived from the requirements document, guide the MNNVL design:
* **Cluster Support:** The feature will only support homogeneous clusters.
* **Global Configuration:** A global configuration for the feature is required. If the cluster does not support MNNVL, the operator must exit.

> * **Opt-Out and Workload Impact:**

Reviewer 3: These are two different things and I would recommended to cover these in 2 different points.
`Opt-Out` - If MNNVL feature is enabled at the operator level then workload operators should be given a provision to opt out of automatic creation of DRA + ComputeDomain resources for specific workloads.

  * An immutable, granular-level opt-out option must be available in the PCS.
  * Failure to create MNNVL resources must not prevent the workload from running.

>   * Changes to the feature configuration should not impact a currently running workload.

Author: fix

* **ComputeDomain Management:**
  * Each PCS replica must have its own dedicated ComputeDomain.
  * The lifecycle of the ComputeDomain is tied directly to the lifecycle of the PCS replica.

# Design Details

## Enabling the feature

Enabling and disabling the feature will be done by the cluster admin by setting a flag in the Grove OperatorConfiguration.

```go
// MNNVLConfiguration defines the configuration for MNNVL (Multi-Node NVLink) support.
type MNNVLConfiguration struct {
   // Enabled indicates whether MNNVL support is enabled.

>    // When true, the operator validates that the ComputeDomain CRD is installed at startup.

Author: taken

   // If validation fails, the operator will exit with an error.
   Enabled bool `json:"enabled"`
}
```

The value could be set from a Helm chart under the config attribute

```yaml
config:
	mnnvl:
		enabled: false
```

> ### Feature validity 

Author: added

When the Grove operator starts, it will check for MNNVL support (only if the feature is enabled). Support is confirmed by the presence of the ComputeDomain CRD in the cluster.

If the cluster lacks MNNVL support, the Grover operator will terminate and log an appropriate error.

> ## ComputeDomain life-cycle management

Reviewer 3: life-cycle - just use a single word lifecycle (its valid)

### Creation

The PodCliqueSet (PCS) controller determines if a ComputeDomain is necessary during reconciliation, based on these criteria:

1. The feature is activated within Grove’s `OperatorConfiguration`.  
2. The PCS has not explicitly opted out of the feature.  
3. At least one contained PodClique (PC) requires a GPU.

> The controller will create a ComputeDomain resource for each PCS replica if one does not already exist. This resource will be named `{pcs-name}-cd-{replica-index}` and will have the PCS set as its owner reference.

Reviewer 4: To remain consistent with existing named resources it should only be `{pcs-name}-{replica-index}` : currently other resources (like podcliques , pcsg, ..) doesn't include the resource type in their name

For example

```yaml
status:
  conditions:
    - type: ComputeDomainsReady
      status: "True"   # All replicas succeeded
      reason: "AllCreated"
      message: "ComputeDomains created for all 3 replicas"
```

```yaml
status:
  conditions:
    - type: ComputeDomainsReady
      status: "False"
      reason: "PartialFailure"
      message: "ComputeDomain creation failed for replicas: [2]. Error: quota exceeded"

```

#### Alt option 1: Manage a condition per replica

```yaml
status:
  conditions:
    - type: ComputeDomainReady-0
      status: "True"
      reason: "Created"
      message: "my-workload-cd-0"
    - type: ComputeDomainReady-1  
      status: "True"
      reason: "Created"
      message: "my-workload-cd-1"
    - type: ComputeDomainReady-2
      status: "False"
      reason: "CreationFailed"
      message: "Failed to create: quota exceeded"
```

Pro:

* Per-replica granularity  
* Easy look up

Cons:

> * Dynamic type names (less common but valid)  

Author: correct added

* Need to clean up conditions on scale-down

#### Alt option 2: Use a desiccated status structure (not recommended)

```go
type PodCliqueSetStatus struct {
    // ... existing fields ...
    
    // ComputeDomainStatuses tracks the status of ComputeDomain per replica.
    // Key is the replica index as string ("0", "1", "2", ...)
    // +optional
    ComputeDomainStatuses map[string]ComputeDomainStatus `json:"computeDomainStatuses,omitempty"`
}

type ComputeDomainStatus struct {
    // Name is the name of the ComputeDomain resource
    Name string `json:"name"`
    // Ready indicates if the ComputeDomain was successfully created
    Ready bool `json:"ready"`
    // Reason provides the status reason
    // +optional
    Reason string `json:"reason,omitempty"`
    // Message provides additional details (e.g., error message)
    // +optional  
    Message string `json:"message,omitempty"`
    // LastUpdated is when this status was last updated
    // +optional
    LastUpdated *metav1.Time `json:"lastUpdated,omitempty"`
}

```

Pro:

* A well-defined structure makes everything clearer.

Cons:

* Ties the feature to Grove.

### Watching the compute domain

> To ensure the PCS controller reconciles when a `ComputeDomain` is deleted or its status changes, a watch must be added in the PCS controller's RegisterWithManager function. 

Author: I removed the "or its status" which was added by mistake

When the controller creates a `ComputeDomain` resource, it should attach the label `grove.io/podcliqueset` with the value `<pcs-name>` to identify which PCS owns it. 

> As part of the registration process, a new watch will be used to garud the `ComputeDomain` created by the controller (using the `grove.io/podcliqueset`label as a filter).

Reviewer 3: ```suggestion
As part of the registration process, a new watch will be added to enqueue events for `ComputeDomain` created by the controller (using the `grove.io/podcliqueset`label as a filter).
```

```yaml
apiVersion: v1beta1
kind: ComputeDomain
metadata:
 labels:
   grove.io/podcliqueset: <pcs-name>
```

### Scale-Out and Scale-In

When scaling out (replicas increased), the subsequent reconciliation process will identify the ComputeDomains missing for the new replica indices and create them using the identical logic as the initial creation.

> When scaling in( replicas decreased), the PCS controller determines which `ComputeDomains` to delete by comparing the existing ones against the new replica count. Any `ComputeDomain` with a replica index equal to or greater than the new count is removed. This process mirrors the "desired versus existing" logic used for managing `PodClique` and `PCSG`.

Reviewer 4: 'in( ' -> 'in ('

The controller lists existing `ComputeDomains` by label selector, computes expected resources from the current spec, and deletes the excess. The `ComputeDomainsReady` condition is updated to reflect the new state after scale operations.

> PCS Deletion  

Reviewer 3: ```suggestion
### PCS Deletion  
```

Author: fix

When a PodCliqueSet is deleted, all ComputeDomain resources are automatically garbage-collected by Kubernetes through the owner reference mechanism. 

Since each ComputeDomain has controllerReference set to the PCS, Kubernetes will cascade-delete them when the PCS is removed. No explicit cleanup logic is required in the controller for this case.

## PCS opt-out

> Users can opt-out of MNNVL support for a specific `PodCliqueSet` using the `grove.io/mnnvl-enabled` annotation. 

Reviewer 5: Where is it stated/stored for a PCS that it has been started with MNNVL enabled? Consider the following scenario:
* Grove is started with MNNVL enabled
* PCS is created (no opt-out label)
* Grove is restarted with MNNVL disabled
Now the PCS should still be reconciled with its CD & RCT (even for new replicas, right?); How does Grove know that this PCS should be handled as an MNNVL-enabled one, even that it is currently running without MNNVL?

One way is to have this mnnvl-enabled label in both cases - false and true - and decide on its (immutable) value in a mutation webhook.

When the global MNNVL feature is enabled in the `OperatorConfiguration`, individual workloads inherit this setting by default. To disable MNNVL for a specific PCS, set the annotation to "`false`".  

When opting out, the operator will not create a `ComputeDomain` for that PCS.

This annotation is validated as immutable by the PCS validation webhook, meaning that once a PCS is created with or without the annotation, it cannot be changed or removed. This ensures consistent behavior throughout the lifecycle of the workload.

## Pod reconciliation

> As described above, before the PCS controller creates a `PodClique` for a replica, it first creates a `ComputeDomain` for the replica. It waits for it to become ready (with a 3-minute timeout).   

Reviewer 3: what is the rational for a 3 min timeout? Is this configurable? I hope the reconciler worker is not blocked waiting for 3 mins.

Author: removed. CD is considered "ok" if it was able be created

# MNNVL Support Design Doc \- Phase 1

# Overview

This design document details the plan to enable the Grove operator to automatically leverage MNNVL for appropriate workloads.

## Abbreviations

| Abbreviation | Full Name | Description |
|--------------|-----------|-------------|
| CD | ComputeDomain | NVIDIA CRD representing a logical GPU fabric spanning multiple nodes |
| RCT | ResourceClaimTemplate | Kubernetes resource template for dynamic resource allocation |
| PCS | PodCliqueSet | Grove CRD that manages a set of PodCliques and PodCliqueScalingGroups |

> | PC | PodClique | Grove CRD representing a group of related pods |

Author: fix

| PCSG | PodCliqueScalingGroup | Grove CRD that manages scaling of PodCliques |

## Motivation

The MNNVL support feature in Grove is guided by three core design principles:

**Simplicity:** Currently, Kubernetes workloads can take advantage of NVIDIA MNNVL accelerated hardware by explicitly adding NVIDIA ComputeDomain APIs to their workloads. With this feature, when applicable, Grove will abstract the MNNVL hardware from users by automatically creating ComputeDomain custom resources, enabling ComputeDomains support to workloads, and managing the lifecycle of the CRs.

> Power users who require custom ComputeDomain or RCT configurations can opt-out of automatic MNNVL management and specify their own `resourceClaims` directly in the pod spec template.

Reviewer 5: AFAIU we decided for opt-in, not opt-out. This should be updated across the whole document, in multiple places.

**Portability:** The Grove Custom Resource Definitions (CRDs)—specifically PodCliqueSet, PodCliqueScalingGroup, and PodClique—are designed to be generic, excluding any direct references to MNNVL, ComputeDomain, or other NVIDIA-specific components.

To maintain portability, all MNNVL-related configuration is kept external, residing in the operator configuration and runtime annotations. This approach allows the same PodCliqueSet manifest to be used without alteration, regardless of whether the target cluster supports MNNVL.

> **Standard Kubernetes behavior:** ComputeDomain lifecycle follows the same reconciliation pattern as other Grove-managed resources. If CD creation fails (due to transient cluster issues), the sync stops and requeues for retry. Persistent failures are surfaced via Kubernetes Events, allowing cluster administrators to investigate.

Reviewer 5: I suggest to surface failures also as part of the PCS status (e.g. using conditions), to make them easily available also for non-admin users.

* **Homogeneous Cluster Support:** The feature will only support clusters with the exact same type of GPUs on all nodes. (AKA homogeneous cluster)
  * Grove does not validate or enforce cluster homogeneity—it is the cluster admin's responsibility to enable this feature only on clusters that meet this requirement. Enabling MNNVL on heterogeneous clusters may result in undefined scheduling behavior.
* **Global Configuration:** A global configuration for the feature is required. If the cluster does not support MNNVL, the Grove Operator will exit with a non-zero exit code.
* **Opt-Out***
  * An immutable, granular-level opt-out option must be available in the PCS.
* **Workload Impact:**
  * Enabling/Disabling of MNNVL feature should not impact a currently running workload.
* **ComputeDomain Management:**
  * Each PCS replica must have its own dedicated ComputeDomain.
  * The lifecycle of the ComputeDomain is tied directly to the lifecycle of the PCS replica.

## Scope and Limitations

This document covers **Phase 1** of MNNVL support in Grove. See GREP-270 for the full requirements.

**Limitations:**

> - **Homogeneous Clusters Only:** This phase supports only homogeneous clusters where all nodes have identical GPU types and NVLink topology. Grove does not validate or enforce cluster homogeneity—it is the cluster administrator's responsibility to enable this feature only on clusters that meet this requirement. Enabling MNNVL on heterogeneous clusters may result in undefined scheduling behavior.

Reviewer 6: Actually I believe in the current design they would be able to workaround by just opting out. I believe the current opt-out plan was to have one flag at the PCS level but we could potentially make it so that you can opt out at more granular levels as well (if this is not already the case)

> - **Homogeneous Clusters Only:** This phase supports only homogeneous clusters where all nodes have identical GPU types and NVLink topology. Grove does not validate or enforce cluster homogeneity—it is the cluster administrator's responsibility to enable this feature only on clusters that meet this requirement. Enabling MNNVL on heterogeneous clusters may result in undefined scheduling behavior.

Reviewer 7: Could you give more detail about the workaround? You mentioned PCS level, do you mean add a flag when customer apply PCS yaml ?

- **PCS-Level Granularity:** The MNNVL feature is applied at the PCS level—it cannot be targeted to individual PCs or PCSGs within a PCS. Either all pods in a replica receive the RCT reference, or none do.

- **No ComputeDomain Customization:** The ComputeDomain and ResourceClaimTemplate configurations are automatically generated by Grove and cannot be customized. Power users who require custom configurations can opt-out of automatic MNNVL management and specify their own `resourceClaims` directly in the pod spec template.

# Design Details

## Enabling the feature

Enabling and disabling the feature will be done by the cluster admin by setting a flag in the Grove OperatorConfiguration.

```go
// MNNVLConfiguration defines the configuration for MNNVL (Multi-Node NVLink) support.
type MNNVLConfiguration struct {
   // Enabled indicates whether MNNVL support is enabled.

>    // When true, the operator validates that the ComputeDomain CRD is installed at startup.

Author: align with the requirement

```yaml
config:
	mnnvl:
		enabled: false
```

> **Note:** Using the `OperatorConfiguration` for feature enablement is chosen for simplicity in Phase 1. However, a plugin-based approach would provide better decoupling between the MNNVL feature and Grove core, and should be considered for future phases.

### Feature validity 

When the Grove operator starts, it will check for MNNVL support (only if the feature is enabled). Support is confirmed by the presence of the ComputeDomain CRD in the cluster.

If the cluster lacks MNNVL support, the Grove operator will terminate and log an appropriate error.

## ComputeDomain lifecycle management

### Creation

The PCS controller has a reconciliation flow for managing resources in a specific order. The `ComputeDomain` component is synced **before** creating PCs and PCSGs, ensuring the CD exists before pods that reference it are created.

Before creating the `CD`, the controller checks the following criteria:

1. The MNNVL feature is enabled in Grove's `OperatorConfiguration`.
2. The PCS has not opted out via the `grove.io/mnnvl-enabled: "false"` annotation.
3. At least one PC in the PCS requires a GPU.

> If all criteria are met, the controller creates a `ComputeDomain` resource for each PCS replica. The CD is named `{pcs-name}-cd-{replica-index}` and references an RCT named `{pcs-name}-rct-{replica-index}`. The PCS is set as the owner reference to enable automatic garbage collection.

Reviewer 5: Naming should be consistent with other grove resources, AFAIK they do not have the kind abbreviation (e.g. "cd" or "rct") as part of their names.

The controller lists existing `ComputeDomains` by label selector, computes expected resources from the current spec, and deletes the excess.

### PCS Deletion  
When a PodCliqueSet is deleted, all ComputeDomain resources are automatically garbage-collected by Kubernetes through the owner reference mechanism. 

Since each ComputeDomain has controllerReference set to the PCS, Kubernetes will cascade-delete them when the PCS is removed. No explicit cleanup logic is required in the controller for this case.

## PCS opt-out

Users can opt-out of MNNVL support for a specific `PodCliqueSet` using the `grove.io/mnnvl-enabled` annotation. 

When the global MNNVL feature is enabled in the `OperatorConfiguration`, individual workloads inherit this setting by default. To disable MNNVL for a specific PCS, set the annotation to "`false`".  

When opting out, the operator will not create a `ComputeDomain` for that PCS.

This annotation is validated as immutable by the PCS validation webhook, meaning that once a PCS is created with or without the annotation, it cannot be changed or removed. This ensures consistent behavior throughout the lifecycle of the workload.

## PC and PCSG Creation

> When the PCS controller creates a PC or PCSG for a replica, it checks whether a CD exists for that replica. If a CD exists, the controller adds an annotation to the PC/PCSG with the RCT name:

Reviewer 8: ```suggestion
When the PCS controller creates a PC or PCSG for a replica, it checks whether a CD exists for that replica using the 'grove.io/podcliqueset' and 'grove.io/replica-index' labels. If a CD exists, the controller adds an annotation to the PC/PCSG with the RCT name:
```

> **Annotation:**

Reviewer 5: The current approach in grove operator is that PCLQ/PCSG sync extracts info on the PCS by referencing it (see `componentutils.GetPodCliqueSet`). Perhaps it's good enough that the MNNVL state is immutable (as already mentioned in the doc), and that RCT names are known, given the consistent naming pattern? Then the annotation on the PC is redundant.

- **Key:** `grove.io/compute-domain-rct`

> - **Value:** The RCT name for the replica (e.g., `my-pcs-rct-0`)

Reviewer 9: Could you clarify the need for this annotation? I expected the RCT would be defined directly within the Pod spec as part of PCLQ

> - **Value:** The RCT name for the replica (e.g., `my-pcs-rct-0`)

Reviewer 9: I’d prefer to keep PCLQ and PCSG agnostic of this feature. Ideally, the implementation should be transparent to them, treating the RCT simply as a standard field within the Pod spec.

Author: adjust

If no CD exists for the replica (feature disabled or PCS opted out), the annotation is not added.

```yaml
apiVersion: grove.io/v1alpha1
kind: PodClique
metadata:
  name: my-pcs-0-worker
  annotations:
    grove.io/compute-domain-rct: "my-pcs-rct-0"  # Present only if CD exists
  labels:
    grove.io/podcliqueset: my-pcs
    grove.io/replica-index: "0"
```

```yaml
apiVersion: grove.io/v1alpha1
kind: PodCliqueScalingGroup
metadata:
  name: my-pcs-0-scaling
  annotations:
    grove.io/compute-domain-rct: "my-pcs-rct-0"  # Present only if CD exists
  labels:
    grove.io/podcliqueset: my-pcs
    grove.io/replica-index: "0"
```

When a PCSG creates PCs, it propagates the `grove.io/compute-domain-rct` annotation to its child PCs.

### Why Use an Annotation?

> A simpler approach would be to have the Pod controller check if the MNNVL support is required at pod creation time and decide whether to add the RCT reference dynamically. However, this approach has a critical flaw:

Reviewer 9: I don't fully agree with the annotation approach as a blanket solution. For PCLQ, since the CRD already includes the Pod spec, the RCT should be part of the CR from the start—this keeps the controller feature-agnostic while ensuring pod consistency. I recognize that PCSG currently lacks a Pod template in its CRD, so we might need to rely on the annotation there for now. However, for PCLQ, the template should be defined upfront so the controller just reconciles it without needing to 'know' about the feature.

A **validating admission webhook** ensures the `grove.io/mnnvl-enabled` annotation is immutable after PCS creation. Any attempt to add, modify, or remove the annotation on an existing PCS is rejected.

**Example:**

```yaml
# User submits PCS without annotation
apiVersion: grove.io/v1alpha1
kind: PodCliqueSet
metadata:
  name: my-pcs
spec:
  replicas: 2
  template:
    cliques:
      - name: worker
        spec:
          podSpec:
            containers:
              - name: train
                resources:
                  limits:
                    nvidia.com/gpu: "8"

# After mutation webhook (if feature enabled):
apiVersion: grove.io/v1alpha1
kind: PodCliqueSet
metadata:
  name: my-pcs
  annotations:
    grove.io/mnnvl-enabled: "true"  # Added by webhook
spec:
  # ... same spec
```

## ComputeDomain lifecycle management

### Creation

The PCS controller has a reconciliation flow for managing resources in a specific order. The `ComputeDomain` component is synced **before** creating PCLQs and PCSGs, ensuring the CD exists before pods that reference it are created.

Before creating the `CD`, the controller checks the `grove.io/mnnvl-enabled` annotation on the PCS:

> - If `grove.io/mnnvl-enabled: "true"` → Create ComputeDomains for each replica

Author: updated

- If `grove.io/mnnvl-enabled: "false"` or annotation is absent → Skip ComputeDomain creation

Since the annotation is set by the mutating webhook at PCS creation time (based on feature enablement and GPU requirements), the controller logic is simplified to a single annotation check.

**Backward Compatibility:** Existing PCS resources created before the MNNVL feature was deployed will not have the `grove.io/mnnvl-enabled` annotation. These workloads will continue to operate without ComputeDomains, even after the feature is enabled globally. To enable MNNVL for an existing workload, the PCS must be deleted and recreated.

> If MNNVL is enabled, the controller creates a `ComputeDomain` resource for each PCS replica. The CD is named `{pcs-name}-{replica-index}` and references an RCT with the same name. The PCS is set as the owner reference to enable automatic garbage collection. The `app.kubernetes.io/part-of` and `grove.io/podcliqueset-replica-index` labels will be used to determine which ComputeDomain is associated with a specific PCS replica. 

Reviewer 5: @shmuel-runai you're not wrong at all. Replicas are considered fully independent of each other and pods from different replicas are not supposed to communicate with each other.

```yaml
apiVersion: resource.nvidia.com/v1beta1
kind: ComputeDomain
metadata:
  name: my-pcs-0
  labels:
    app.kubernetes.io/managed-by: grove
    app.kubernetes.io/part-of: my-pcs
    app.kubernetes.io/component: pcs-computedomain
    grove.io/podcliqueset-replica-index: "0"
  ownerReferences:
    - apiVersion: grove.io/v1alpha1
      kind: PodCliqueSet
      name: my-pcs
      controller: true
spec:
  channel:
    resourceClaimTemplateName: my-pcs-0
```

### Observability

ComputeDomain creation follows the same observability pattern as other Grove-managed resources:

- **Kubernetes Events:** Success and failure events are emitted on the PCS resource.
  ```
  kubectl describe pcs my-pcs
  # Events:
  #   Normal   ComputeDomainCreated   ComputeDomain my-pcs-0 created
  #   Warning  ComputeDomainFailed    Failed to create ComputeDomain for replica 2: <error>
  ```

- **Logs:** Detailed error information is logged by the operator.

- **Requeue on failure:** If CD creation fails, the sync stops and requeues for retry. PCLQ and PCSG creation only proceeds after all CDs for the replicas have been successfully created.

### Protecting the ComputeDomain

> Deleting a ComputeDomain while pods are actively using its RCT causes significant workload disruption. To prevent accidental deletion, Grove adds a **finalizer** to each ComputeDomain it creates.

Reviewer 9: Did you checked this? It might not required  if CT does it by it own

If the cluster lacks MNNVL support, the Grove operator will terminate and log an appropriate error.

### PCS MNNVL Eligibility Determination

When a PodCliqueSet is created, a **mutating admission webhook** automatically determines and records the MNNVL enablement status by setting the `grove.io/mnnvl-enabled` annotation.

The webhook behavior:

1. **If annotation already exists** → Do not override (respects user opt-out)
2. **If MNNVL feature is disabled** in `OperatorConfiguration` → Do not add annotation
3. **If MNNVL feature is enabled**:
   - PCS has GPU containers (`nvidia.com/gpu`) → Set `grove.io/mnnvl-enabled: "true"`

>    - PCS has no GPU containers → Set `grove.io/mnnvl-enabled: "false"`

Reviewer 5: I prefer that we will never add a false annotation. We discussed it and understood we don't need it. The only case where we should see a PCS with a false annotation is when the user added it in order to flag us to opt-out.

Author: updated

**Opt-out Behavior:**

If the annotation already exists on the PCS (user explicitly set it), the webhook does **not** override it. Users can opt-out of MNNVL support for a specific `PodCliqueSet` by explicitly setting `grove.io/mnnvl-enabled: "false"` in the PCS manifest **before creation**.

> When a PCS is submitted with this annotation already set, the mutating webhook will **not override** the user's choice. This allows workloads to explicitly disable MNNVL even when the feature is globally enabled and the workload requires GPUs.

Reviewer 5: There's one more case to talk about - either here or later in this section. If the user sets the annotation to true, but mnnvl is disabled by the operator. It should fail the submission. For existing workloads it's okay, but not for new ones.

## PCLQ Creation and RCT Injection

The `resourceClaims` reference is injected into the PCLQ's pod spec template at **PCLQ creation time**, not at pod creation time. This ensures the decision is made once and baked into the PCLQ spec.

### PCS Creating PCLQ

When the PCS controller creates a PCLQ, it checks:
1. Does the PCS have `grove.io/mnnvl-enabled: "true"`?
2. Does the PCLQ's pod spec require GPU (`nvidia.com/gpu`)?

If both conditions are true, the controller injects `resourceClaims` into the PCLQ's pod spec template before creation:

```yaml
apiVersion: grove.io/v1alpha1
kind: PodClique
metadata:
  name: my-pcs-0-worker
  labels:
    app.kubernetes.io/part-of: my-pcs
    grove.io/podcliqueset-replica-index: "0"
spec:
  podSpec:
    resourceClaims:
      - name: mnnvl-claim
        resourceClaimTemplateName: my-pcs-0  # {pcs-name}-{replica-index}
    containers:
      - name: train
        image: my-training-image
        resources:
          limits:
            nvidia.com/gpu: "8"
          claims:
            - name: mnnvl-claim
```

### PCS Creating PCSG

> When the PCS controller creates a PCSG, it propagates the `grove.io/mnnvl-enabled` annotation to the PCSG:

Reviewer 5: Please consider not adding this annotation on the PCSG but read it from the PCS (when syncing the PCSG) - as many other attributes are already read from the PCS at exactly the same stage in the controller's reconcile loop.
